tests/integration/test_containerd_integration.py

```diff
@@ -79,6 +79,38 @@ async def test_upgrade_dry_run_action(ops_test, which_action):
     assert end >= start, "containerd service shouldn't have been restarted"
 
 
+async def test_upgrade_action_containerd_force(ops_test):
+    """Test running upgrade action without GPU and with force."""
+    unit = ops_test.model.applications["containerd"].units[0]
+    start = await process_elapsed_time(unit, "containerd")
+    action = await unit.run_action("upgrade-packages", **{"force": True})
+    output = await action.wait()  # wait for result
+    assert output.data.get("status") == "completed"
+    results = output.data.get("results", {})
+    log.info(f"Upgrade results = '{results}'")
+    assert results["containerd"]["available"] == results["containerd"]["installed"]
+    assert results["containerd"]["upgrade-available"] == "False"
+    assert not results["containerd"].get("upgrade-completed"), "No upgrade should have been run"
+    end = await process_elapsed_time(unit, "containerd")
+    assert end >= start, "containerd service shouldn't have been restarted"
+
+
+async def test_upgrade_action_without_gpu_gpu_force(ops_test):
+    """Test running upgrade action with GPU and force without GPU does nothing."""
```

Comment:
Perhaps also an enlightening comment.  It'd need to be run though black.

```suggestion
    """Test running upgrade action with no GPU drivers installed.
    
        action with the `GPU` and `force` flags both set
        but without the GPU drivers installed should not upgrade any gpu drivers."""
```